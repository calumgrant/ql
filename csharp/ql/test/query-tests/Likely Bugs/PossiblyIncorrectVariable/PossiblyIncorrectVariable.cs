using System;

namespace Test
{
    class Point1
    {
        public double x, y;

        public Point1(double x, double y)
        {
            this.x = x;
            this.x = y;  // this.x should be this.y.
        }

        public Point1(double x, double y, double z)
        {
            this.x = x;
            this.y = y;  // Good
        }
    }

    class Point2
    {
        public double mX { get; set; }
        public double mY { get; set; }

        public Point2(double x, double y)
        {
            mX = x;
            mY = x;  // x should be y.
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            var p1 = new Point1(1, 2);
            var p2 = new Point1(1, 2);
            var p3 = new Point2(1, 2);
            var p4 = new Point2(1, 2);
            double m_x, m_y;

            if (p1.x == p1.x &&
                p1.x == p2.y)  // x should be y
            {
            }

            if (p1.x == p1.x &&
                p1.y == p2.y)  // Good
            {
            }

            p2.x = p1.x;
            p2.y = p1.x;  // x should be y

            p1.x = p2.x;
            p1.x = p2.y;  // x should be y

            p3.mX = p1.x;
            p3.mX = p1.y;  // mX should be mY

            p1.x = p3.mX;
            p1.x = p3.mY;  // x should be y

            m_x = p1.x;
            m_x = p1.y;  // x should be y

            m_x = p1.x;
            m_y = p1.y;  // Good

            var x = p4.mX;
            var y = p4.mX;  // mX should be mY

            double Mx = p1.x, My = p1.x;  // x should be y

            double Nx = p1.x, Ny = p1.y;  // Good

            double NN1x = p1.x, NN2y = p1.x;  // Good - Nx and NNy are sufficiently unrelated

            if (p1.x != p1.x ||
                p1.x != p2.y)  // x should be y
            {
            }

            if (p1.x != p1.x ||
                p1.y != p2.y)  // Good
            {
            }

            {
                // Tests block scope
                double Lx = 1, Ly = 2;
                p2.x = Lx;
                p2.x = Ly;  // x should be y

                p2.x = Lx;
                p2.y = Lx;  // Lx should be Ly
            }
            {
                // Tests block scope
                double Lx = 1, Ly = 2;
                p2.x = Lx;
                p2.x = Ly;  // x should be y

                p2.x = Lx;
                p2.y = Lx;  // Lx should be Ly
            }
        }
    }
}
