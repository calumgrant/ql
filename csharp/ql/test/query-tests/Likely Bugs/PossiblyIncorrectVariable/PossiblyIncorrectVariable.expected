| PossiblyIncorrectVariable.cs:12:13:12:18 | access to field x | Duplicated variable 'x' may be an error, because '$@' might have been intended. | PossiblyIncorrectVariable.cs:7:26:7:26 | y | y |
| PossiblyIncorrectVariable.cs:30:18:30:18 | access to parameter x | Duplicated variable 'x' may be an error, because '$@' might have been intended. | PossiblyIncorrectVariable.cs:27:40:27:40 | y | y |
| PossiblyIncorrectVariable.cs:45:17:45:20 | access to field x | Duplicated variable 'x' may be an error, because '$@' might have been intended. | PossiblyIncorrectVariable.cs:7:26:7:26 | y | y |
| PossiblyIncorrectVariable.cs:55:20:55:23 | access to field x | Duplicated variable 'x' may be an error, because '$@' might have been intended. | PossiblyIncorrectVariable.cs:7:26:7:26 | y | y |
| PossiblyIncorrectVariable.cs:58:13:58:16 | access to field x | Duplicated variable 'x' may be an error, because '$@' might have been intended. | PossiblyIncorrectVariable.cs:7:26:7:26 | y | y |
| PossiblyIncorrectVariable.cs:61:13:61:17 | access to property mX | Duplicated variable 'mX' may be an error, because '$@' might have been intended. | PossiblyIncorrectVariable.cs:25:23:25:24 | mY | mY |
| PossiblyIncorrectVariable.cs:64:13:64:16 | access to field x | Duplicated variable 'x' may be an error, because '$@' might have been intended. | PossiblyIncorrectVariable.cs:7:26:7:26 | y | y |
| PossiblyIncorrectVariable.cs:67:13:67:15 | access to local variable m_x | Duplicated variable 'm_x' may be an error, because '$@' might have been intended. | PossiblyIncorrectVariable.cs:42:25:42:27 | m_y | m_y |
| PossiblyIncorrectVariable.cs:73:21:73:25 | access to property mX | Duplicated variable 'mX' may be an error, because '$@' might have been intended. | PossiblyIncorrectVariable.cs:25:23:25:24 | mY | mY |
| PossiblyIncorrectVariable.cs:75:36:75:39 | access to field x | Duplicated variable 'x' may be an error, because '$@' might have been intended. | PossiblyIncorrectVariable.cs:7:26:7:26 | y | y |
| PossiblyIncorrectVariable.cs:82:17:82:20 | access to field x | Duplicated variable 'x' may be an error, because '$@' might have been intended. | PossiblyIncorrectVariable.cs:7:26:7:26 | y | y |
| PossiblyIncorrectVariable.cs:95:17:95:20 | access to field x | Duplicated variable 'x' may be an error, because '$@' might have been intended. | PossiblyIncorrectVariable.cs:7:26:7:26 | y | y |
| PossiblyIncorrectVariable.cs:98:24:98:25 | access to local variable Lx | Duplicated variable 'Lx' may be an error, because '$@' might have been intended. | PossiblyIncorrectVariable.cs:93:32:93:33 | Ly | Ly |
| PossiblyIncorrectVariable.cs:104:17:104:20 | access to field x | Duplicated variable 'x' may be an error, because '$@' might have been intended. | PossiblyIncorrectVariable.cs:7:26:7:26 | y | y |
| PossiblyIncorrectVariable.cs:107:24:107:25 | access to local variable Lx | Duplicated variable 'Lx' may be an error, because '$@' might have been intended. | PossiblyIncorrectVariable.cs:102:32:102:33 | Ly | Ly |
| PossiblyIncorrectVariableBad.cs:10:9:10:9 | access to property X | Duplicated variable 'X' may be an error, because '$@' might have been intended. | PossiblyIncorrectVariableBad.cs:5:19:5:19 | Y | Y |
